### PR TITLE
New Error Handling Pattern

### PR DESCRIPTION
**Pull request by gshaw1997**
### **Motivation**
Services currently use `try catch` blocks to handle all errors that may occur during asynchronous processes. At the driver level, any errors that are thrown are caught and returned as rejected Promises instead of thrown. There is also no status code associated with errors that occur. This makes it difficult to track down issues in services as well as send the appropriate status code to the client and message.

### **Proposed Solution**
To remedy this problem, I suggest that we remove rejected Promises from the driver level and instead throw the appropriate `Error` object. We should also use `console.error` at the interactor level to log out stack traces from thrown errors. Lastly we should return a specialized error object to the client that contains information needed to diagnose the issue and display a user facing message. 

To facilitate these changes, I have created four `Reason` types, two new `Error` objects, `DriverError` and `ServiceError`, an interface `ClientError`, and a `handlePromise` function. These components work together to create consistency in the way errors are handled and to make errors more descriptive and meaningful.

### **Reason Enums & Types**
The reason enums are used to map the cause of an error to an appropriate response status code. `Reason` types are used to extend the various reason enums.
##### `REASON`- Can be used in any `Error` Object
- `BAD_REQUEST`: maps to status code `400`. To be used in the case where a failure occurs due to request parameters being malformed.
- `METHOD_NOT_IMPLEMENTED`: maps to status code `501`. To be used in the case where a failure occurs due to a method not being implemented.
##### `GENERIC_REASON`- Can be used in any `Error` Object
- `UNEXPECTED_ERROR`: maps to status code `500`. To be used in the case where a failure occurs due to an unknown issue.
##### `DRIVER_REASON` - Can be used in `DriverError` object 
- `RESOURCE_NOT_FOUND`: maps to status code `404`. To be used in the case where a failure occurs due to an unresolvable resource .
##### `SERVICE_REASON` - Can be used in `ServiceError` object
- `NOT_AUTHORIZED`: maps to status code `403`. To be used in the case where a failure occurs because of user's access privileges.
#### Types
- `GenericReason`: is equal to the `GENERIC_REASON` enum.
- `DriverReason`: is the union type between the `GenericReason` type, `DRIVER_REASON` enum, and the `REASON` enum.
- `ServiceReason`: is the union type between the `GenericReason` type, `SERVICE_REASON` enum, and the `REASON` enum.
- `Reason`: is the union type of the `REASON` enum, `GenericReason` type, `DriverReason`, and `ServiceReason`.
### **Error Objects**
- `DriverError`: is to be thrown at the driver level and provided a `message:string` and `reason:DriverReason`
- `ServiceError`: is to be thrown at the service level (any level external to driver level as of now) and provided a `message:string` and `reason:ServiceReason`
### **ClientError**
This interface defines the structure of the object to be sent to the client when an error occurs.
### **handlePromise**
This function wraps Promises,  behaves like a try catch block, and can be passed in an `Error` object. If the Promise is successfully resolved the data is returned as expected. If an error occurs it catches the error and applies custom error handling logic. If an `Error` object is provided, the catch block will inject the message from the caught error if no custom message is provided and throw the `Error` object.

This function allows error to be handled at a more granular level than a `try catch` block, which can encourage more descriptive error messages and usage of appropriate status codes.

### CODE REVIEW

File: src/interactors/SuggestionInteractor.ts [hunk 1]
```diff
@@ -82,9 +92,15 @@ export class SuggestionInteractor {
    */
   public static async fetchSources(dataStore: DataStore): Promise<string[]> {
     try {
-      return dataStore.fetchSources();
+      return await dataStore.fetchSources();
     } catch (e) {
-      return Promise.reject(`Problem finding sources. Error: ${e}.`);
+      console.error(e);
```
Comment (nrwinner) [inline src/interactors/SuggestionInteractor.ts]: This can be removed since handlePromise logs errors

File: src/interactors/SuggestionInteractor.ts [hunk 2]
```diff
@@ -58,17 +64,21 @@ export class SuggestionInteractor {
...
-  ): Promise<void> {
+  ): Promise<{ total: number; outcomes: StandardOutcomeDocument[] }> {
     try {
       filter = this.sanitizeFilter(filter);
-      const suggestions = await dataStore.searchOutcomes(filter, limit, page);
-      responder.sendObject(suggestions);
+      return await dataStore.searchOutcomes(filter, limit, page);
     } catch (e) {
-      responder.sendOperationError(`Problem searching outcomes. Error: ${e}.`);
+      console.error(e);
```
Comment (nrwinner) [inline src/interactors/SuggestionInteractor.ts]: This can be removed since handlePromise logs errors

File: src/interactors/SuggestionInteractor.ts [hunk 3]
```diff
@@ -21,27 +23,31 @@ export class SuggestionInteractor {
...
         filter,
         mode,
         threshold,
         limit,
         page,
       );
-      responder.sendObject(suggestions);
     } catch (e) {
-      responder.sendOperationError(`Problem suggesting outcomes. Error: ${e}.`);
+      console.error(e);
```
Comment (nrwinner) [inline src/interactors/SuggestionInteractor.ts]: This can be removed since handlePromise logs errors

File: src/interactors/SuggestionInteractor.ts [hunk 4]
```diff
@@ -21,27 +23,31 @@ export class SuggestionInteractor {
...
+      const clientErr: ClientError = {
```
Comment (DonSeannelly) [inline src/interactors/SuggestionInteractor.ts]: HTTP status is a detail that an Interactor cannot be aware of within its architectural bounds. Could we perhaps push this discernment back to the RouteHandler that will catch this promise rejection?

**Comment by DonSeannelly** [comment]
I don't think that I fully wrap my head around the need for the `handlePromise` function yet. Could we simply reject one of these new types (in the case of a driver, just reject a `DriverError` inside of the catch statement)? With this implementation every time that a function is called it is making and passing around an error object, which seems to me to be a detail out of scope of a successful run.

TL;DR `handleError` seems verbose, can we just reject a `DriverError` inside the typical catch block.

File: src/types/reason.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
+export enum REASON {
+  BAD_REQUEST = 400,
+  METHOD_NOT_IMPLEMENTED = 501,
+}
+export enum GENERIC_REASON {
```
[the remark this answers is not in the document]
Comment (gshaw1997) [reply src/types/reason.ts]: It was segmented because it is defined as ["The general catch-all error when the server-side throws an exception."](https://www.restapitutorial.com/httpstatuscodes.html#) , therefore being generic, whereas the other codes cover more specific error cases.

File: src/interactors/SuggestionInteractor.ts (hunk 4 above)
Comment (gshaw1997) [reply src/interactors/SuggestionInteractor.ts]: Resolved here 5aa95e85e3220487311240d5046390a3fd4db278.

File: src/types/DriverError.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import { DriverReason } from './reason';
+
+export class DriverError extends Error {
+  reason: DriverReason;
+
+  constructor(message: string, reason: DriverReason) {
+    super(message);
+    this.reason = reason;
+    // Set the prototype explicitly.
+    Object.setPrototypeOf(this, DriverError.prototype);
```
Comment (DonSeannelly) [inline src/types/DriverError.ts]: Why do we have to do this?

File: src/types/reason.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (DonSeannelly) [reply src/types/reason.ts]: Will there ever be more than one generic reason? Or will a generic reason always be an unexpected error?

File: src/types/DriverError.ts (hunk 6 above)
Comment (gshaw1997) [reply src/types/DriverError.ts]: This is needed in order to override Javascript's `Error` class's `toString` method.

File: src/types/reason.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (gshaw1997) [reply src/types/reason.ts]: Not that I thought of

**Comment by gshaw1997** [comment]
This PR sparked great discussion and led to the standardization of another error handling pattern which we are implementing throughout our services. 

It can be see in practice here

https://github.com/Cyber4All/rating-service/pull/16/commits/8139d3d1f0fa430220565eb7680794cf3cf37744

https://github.com/Cyber4All/rating-service/pull/16/files#diff-95f55f0fd049cf453de496dca246a2b5
